scilpy/io/streamlines.py

```diff
@@ -144,3 +144,33 @@ def load_tractogram_with_reference(parser, args, filepath,
         parser.error('{} is an unsupported file format'.format(filepath))
 
     return sft
+
+
+def filter_tractogram_data(tractogram, streamline_ids):
+    """ Filter tractogram data (data_per_streamline, data_per_point)
+    according to streamline ids
+
+    Parameters:
+    -----------
+    tractogram: Tractogram or StatefulTractogram
+        Tractogram containing the data to be filtered
+    streamline_ids: array_like
+        List of streamline ids the data corresponds to
+
+    Returns:
+    --------
+    A tuple containing
+        Dict of filtered data_per_streamline
+        Dict of filtered data_per_point
+    """
+
+    new_data_per_streamline = {}
+    new_data_per_point = {}
+    for key in tractogram.data_per_streamline.keys():
```

Comment:
Without doing a for loop, there is code in scil_outlier_rejection.py showing it.

```
outliers, inliers = prune(streamlines, args.alpha, summary)
inliers_streamlines = tractogram.streamlines[inliers]
inliers_data_per_streamline = tractogram.tractogram.data_per_streamline[inliers]
inliers_data_per_point = tractogram.tractogram.data_per_point[inliers]

outliers_streamlines = tractogram.streamlines[outliers]
outliers_data_per_streamline = tractogram.tractogram.data_per_streamline[outliers]
outliers_data_per_point = tractogram.tractogram.data_per_point[outliers]
```

Reply:
Ahhh I didn't know about this, will do !

<rant>
Holy crap that's some (damn useful) black magic right there. I had to test it myself and then go look at `nibabel.streamlines.tractogram.PerArrayDict`'s class definition to see what's the intended behavior.

For anyone else coming across this discussion who had no clue about this like I did:
```
class PerArrayDict(SliceableDataDict):
    r""" Dictionary for which key access can do slicing on the values.
    This container behaves like a standard dictionary but extends key access to
    allow keys for key access to be indices slicing into the contained ndarray
    values. The elements must also be ndarrays.
    In addition, it makes sure the amount of data contained in those ndarrays
    matches the number of streamlines given at the instantiation of this
    instance.
```
</rant>